experimental/vjn/yarpl/include/reactivestreams/subscription.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "subscriber.h"
+
+namespace reactivestreams {
+
+template <typename T>
+class Subscription {
+ public:
+  virtual ~Subscription() = default;
+  virtual void request(long n) = 0;
+  virtual void cancel() = 0;
+};
+
+template <typename T, typename Next>
+class DeletingSubscription : public Subscription<T> {
+ public:
+  DeletingSubscription(std::shared_ptr<Next> function,
+                       std::unique_ptr<Subscriber<T>> subscriber)
+      : function_(function), subscriber_(std::move(subscriber)) {}
+
+  void start() {
+    subscriber_->on_subscribe(this);
+
+    try {
+      (*function_)(*subscriber_);
+      subscriber_->on_complete();
+    } catch (...) {
+      subscriber_->on_error(std::current_exception());
+    }
+
+    delete this;
```

Comment:
with this memory management model, however with good intentions, we will end up in a SEGFAULT hell. 
We should let shared_ptrs and unique_ptrs do the memory management job. There should be no new or delete in the code (and never "delete this").